cmd/snap-update-ns/bootstrap.c

```diff
@@ -136,7 +144,13 @@ static int switch_to_privileged_user()
 	}
 	// After changing uid, our effective capabilities were dropped.
 	// Reacquire CAP_SYS_ADMIN, and discard CAP_SETUID/CAP_SETGID.
-	data[0].effective = CAP_TO_MASK(CAP_SYS_ADMIN);
+	data[0].effective = (
+		/* To perform mount operations. */
+		CAP_TO_MASK(CAP_SYS_ADMIN) |
+		/* To ignore filesystem permissions. */
+		CAP_TO_MASK(CAP_DAC_OVERRIDE) |
+		CAP_TO_MASK(CAP_DAC_READ_SEARCH)
```

Comment:
I really dislike keeping this for the duration of snap-update-ns because it severely weakens the benefit of dropping privileges. Between SYS_ADMIN and these, snap-update-ns can do so much... (though on systems with apparmor, snap-update-ns is quite constrained wrt writes by a snap-specific apparmor profile, but on systems without apparmor we don't).

Needing DAC_OVERRIDE and DAC_READ_SEARCH typically indicates a design flaw with the privilege dropping and should be fixable without requiring keeping the capability. Is it not possible to open the fd prior to dropping privileges here in the bootstrap code and then passing it to snap-update-ns? I guess part of the problem with that is that then the bootstrap code might have to parse things.... What if snap-update-ns instead works on a temporary file and then after it returns the calling process puts it into place?

Considering the snap-update-ns apparmor profile, I'm grudgingly willing to accept this as is, but not without understanding why the couple of ideas I had (and any others you had) for avoiding it are worse than keeping the caps. We may want to include those as a code comment, but let's see what happens with the discussion.

Btw, if you have DAC_OVERRIDE, you probably shouldn't need DAC_READ_SEARCH; did you verify you need both?